tle/cogs/acd_ai.py

```diff
@@ -0,0 +1,185 @@
+from discord.ext import commands
+from collections.abc import AsyncIterable
+import discord
+import google.generativeai as genai
+from tle.util import settings
+import requests
+from PIL import Image
+from tle import constants
+from ratelimit import limits, RateLimitException, sleep_and_retry
+
+
+
+ONE_MINUTE = 60
+MAX_CALLS_PER_MINUTE = 60
+API_KEY = constants.API_KEY
+
+image_types = ["jpeg", "png", "webp", "heic", "heif"]
+no_response_message = "Sorry, can't answer that. possible reasons might be no response, recitation, safety issue or blocked content."
+help_text = "```\nUsage:\n$chat : to create a public thread to interact with the bot.\n$private : to create a private thread to interact with the bot.\n\nACD AI answers to all text, image and attachment.\nHave fun!\n\nfor Admin/Mods:\n$set : to set a channel.\n$unset : to unset a channel.\n$show : to show current channels.\n$clear : to delete all the threads in set channels.```"
+allowed_guilds = [1063393625049923687, 501032525284769842]
+dev_id = 501026469569363988
+
+class ACDAI(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.allowed_channels = []
+        self.allowed_guilds = allowed_guilds
+        self.dev_id = dev_id
+        genai.configure(api_key=API_KEY)
+        self.text_model = genai.GenerativeModel(model_name="gemini-pro",
+                              generation_config=settings.text_generation_config, # type: ignore
+                              safety_settings=settings.text_safety_settings)
+        self.image_model = genai.GenerativeModel(model_name="gemini-pro-vision",
+                              generation_config=settings.image_generation_config, # type: ignore
+                              safety_settings=settings.image_safety_settings)
+
+        self.chats = {}
+    
+    async def print_response(self, response, message):
+        async with message.channel.typing():
+            strings = []
+            string = []
+            count = 0
+            for char in response:
+                if count < 2000: string.append(char); count += 1
+                else: strings.append("".join(string)); string = [char]; count = 1
+            strings.append("".join(string))
+            try:
+                await message.reply(strings[0])
+                for string in strings[1:]:
+                    await message.channel.send(string)
+            except Exception as e:
+                message.reply("No response, discord issue :(")
+                print("last message:", message.content)
+                print(e)
```

Comment:
can be merged in a single line for more readability.

Reply:
Will fix.